Core/Src/ACDC_SPI.c

```diff
@@ -0,0 +1,96 @@
+/**
+ * @file ACDC_SPI.h
+ * @author Liam Bramley
+ * @brief Implementation of SPI driver
+ * 
+ * This file includes functions to initalize the SPI, transfer and recieve data.
+ * 
+ * @version 0.1
+ * @date 2024-01-30
+ * @copyright Copyright (c) 2023-2024
+ * 
+ */
+
+#include "ACDC_SPI.h"
+
+
+/*
+{
+  __IO uint32_t CR1;      /* SPI control register 1         */
+  __IO uint32_t CR2;      /* SPI control register 2         */
+  __IO uint32_t SR;       /* SPI status register            */
+  __IO uint32_t DR;       /* SPI data register              */
+  __IO uint32_t CRCPR;    /* SPI CRC polynomial register    */
+  __IO uint32_t RXCRCR;   /* SPI RX CRC register            */
+  __IO uint32_t TXCRCR;   /* SPI TX CRC register            */
+  __IO uint32_t I2SCFGR;  /* SPI_I2S configuration register 
+ SPI_TypeDef;
+*/
+
+void SPI_Initalize(const SPI_TypeDef *SPI) {
+    if(SPI == SPI1){
+        // Enable clock for SPI
+        RCC->APB2ENR |= RCC_APB2ENR_SPI1EN;
+
+        // Configure GPIO pins for SPI1
+        // Set PA5 (SCK) as alternate function push-pull output
+        GPIO_PinDirection(GPIOA, GPIO_PIN_5, GPIO_MODE_OUTPUT_SPEED_50MHz, GPIO_CNF_OUTPUT_AF_PUSH_PULL);
+
+        // Set PA6 (MISO) as input floating
+        GPIO_PinDirection(GPIOA, GPIO_PIN_6, GPIO_MODE_INPUT, GPIO_CNF_INPUT_FLOATING);
+
+        // Set PA7 (MOSI) as alternate function push-pull output
+        GPIO_PinDirection(GPIOA, GPIO_PIN_7, GPIO_MODE_OUTPUT_SPEED_50MHz, GPIO_CNF_OUTPUT_AF_PUSH_PULL);
+
+        // Configure SPI1
+        SPI1->CR1 = 0; // Clear CR1 register
+        SPI1->CR1 |= SPI_CR1_SSM | SPI_CR1_SSI | SPI_CR1_BR_0 | SPI_CR1_BR_1; // Set master mode, software NSS, baud rate
```

Comment:
Looks like the Baud rate is locked to the Frequency of the PClk/4. May want a way to change this... {See RM-743}

Here is some example code you could use 
```c
typedef enum{ // SPI Baud Rate Divider
    SPI_BAUD_DIV_2   = 0b000,   /**< Baud rate: SysClk / 2 bps   */
    SPI_BAUD_DIV_4   = 0b001,   /**< Baud rate: SysClk / 4 bps   */
    SPI_BAUD_DIV_8   = 0b010,   /**< Baud rate: SysClk / 8 bps   */
    SPI_BAUD_DIV_16  = 0b011,   /**< Baud rate: SysClk / 16 bps  */
    SPI_BAUD_DIV_32  = 0b100,   /**< Baud rate: SysClk / 32 bps  */
    SPI_BAUD_DIV_64  = 0b101,   /**< Baud rate: SysClk / 64 bps  */
    SPI_BAUD_DIV_128 = 0b110,   /**< Baud rate: SysClk / 128 bps */
    SPI_BAUD_DIV_256 = 0b111    /**< Baud rate: SysClk / 256 bps */
}SPI_BaudDivider;

void SPI_SetBaudDivider(SPI_TypeDef *SPIx, SPI_BaudDivider SPI_BAUD_DIV_x){   
    CLEAR_BIT(SPIx->CR1, SPI_CR1_BR_Msk);                   // Clear the SPIx Baud Divisor bits
    SET_BIT(SPIx->CR1, SPI_BAUD_DIV_x << SPI_CR1_BR_Pos);   // Set the SPIx Baud Divisor bits
}
```